File: src/service/port.ts
/**
 * Service port interfaces — the boundary between UI/CLI and game logic.
 *
 * ServicePort is the production interface. DevServicePort extends it with
 * debug/observability methods. Production stores type against ServicePort;
 * debug panels cast to DevServicePort.
 */

import type { Call, Card, Seat } from "../engine/types";
import type { PlayProfileId } from "./session-types";
import type {
  DrillHandle,
  SessionConfig,
} from "./request-types";
import type {
  BiddingViewport,
  DeclarerPromptViewport,
  PlayingViewport,
  ExplanationViewport,
  ModuleCatalogEntry,
  ModuleLearningViewport,
  DrillStartResult,
  BidSubmitResult,
  PlayEntryResult,
  PlayCardResult,
  DDSolutionResult,
  ConventionInfo,
  BundleFlowTreeViewport,
  ModuleFlowTreeViewport,
  ServiceInferenceSnapshot,
  ServicePublicBeliefState,
} from "./response-types";
import type {
  ServiceDebugLogEntry,
} from "./debug-types";

/** Production service interface — all methods return Promise<T>. */
interface ServicePort {
  /** Bootstrap the service. Transport-specific wiring (WASM load, DDS, etc.) happens here.
   *  Must be called once before any other method. Idempotent. */
  init(): Promise<void>;

  // ── Session lifecycle ───────────────────────────────────────────
  createDrillSession(config: SessionConfig): Promise<DrillHandle>;

  // ── Drill lifecycle ─────────────────────────────────────────────
  startDrill(handle: DrillHandle): Promise<DrillStartResult>;

  // ── Bidding ─────────────────────────────────────────────────────
  /** Grade + apply + run AI + return next viewport — single round-trip. */
  submitBid(handle: DrillHandle, call: Call): Promise<BidSubmitResult>;

  // ── Phase transitions ───────────────────────────────────────────
  enterPlay(handle: DrillHandle, seatOverride?: Seat): Promise<PlayEntryResult>;
  declinePlay(handle: DrillHandle): Promise<void>;
  returnToPrompt(handle: DrillHandle): Promise<void>;
  restartPlay(handle: DrillHandle): Promise<PlayEntryResult>;

  // ── Play ────────────────────────────────────────────────────────
  playCard(handle: DrillHandle, card: Card, seat: Seat): Promise<PlayCardResult>;
  skipToReview(handle: DrillHandle): Promise<void>;
  updatePlayProfile(handle: DrillHandle, profileId: PlayProfileId): Promise<void>;

  // ── Query ───────────────────────────────────────────────────────
  getBiddingViewport(handle: DrillHandle): Promise<BiddingViewport | null>;
  getDeclarerPromptViewport(handle: DrillHandle): Promise<DeclarerPromptViewport | null>;
  getPlayingViewport(handle: DrillHandle): Promise<PlayingViewport | null>;
  getExplanationViewport(handle: DrillHandle): Promise<ExplanationViewport | null>;

  // ── Inference ─────────────────────────────────────────────────────
  /** Get the current public belief state from the session's inference coordinator.
   *  Eliminates the need for the store to maintain its own inference coordinator. */
  getPublicBeliefState(handle: DrillHandle): Promise<ServicePublicBeliefState>;

  // ── DDS analysis ────────────────────────────────────────────────
  getDDSSolution(handle: DrillHandle): Promise<DDSolutionResult>;

  // ── Convention catalog ──────────────────────────────────────────
  listConventions(): Promise<ConventionInfo[]>;

  // ── Learning ──────────────────────────────────────────────────
  /** List all registered modules with catalog metadata. */
  listModules(): Promise<readonly ModuleCatalogEntry[]>;
  /** Build a learning viewport for a single module by ID. */
  getModuleLearningViewport(moduleId: string): Promise<ModuleLearningViewport | null>;

  /** Build a unified conversation flow tree for a bundle. */
  getBundleFlowTree(bundleId: string): Promise<BundleFlowTreeViewport | null>;

  /** Build a conversation flow tree scoped to a single module. */
  getModuleFlowTree(moduleId: string): Promise<ModuleFlowTreeViewport | null>;
}

/** Extends ServicePort with dev/debug methods.
 *  Production builds use ServicePort only.
 *  local-service.ts implements both. */
export interface DevServicePort extends ServicePort {
  getExpectedBid(handle: DrillHandle): Promise<{ call: Call } | null>;
  getDebugLog(handle: DrillHandle): Promise<readonly ServiceDebugLogEntry[]>;
  getInferenceTimeline(handle: DrillHandle): Promise<readonly ServiceInferenceSnapshot[]>;

  /** Return the resolved convention name for a session. */
  getConventionName(handle: DrillHandle): Promise<string>;

  /** Create a session from a pre-built DrillBundle (for tests using stub engines).
   *  @deprecated Only available in LocalService (TS backend). Not supported in WasmService. */
  createDrillSessionFromBundle(bundle: unknown): Promise<DrillHandle>;
}

File: src-tauri/crates/bridge-service/src/port.rs
//! ServicePort and DevServicePort traits — the hexagonal boundary.
//!
//! All methods are synchronous. The UI/WASM layer calls these traits;
//! implementations delegate to bridge-session controllers.

use bridge_engine::types::{Call, Card, Seat};
use bridge_session::session::{
    BiddingViewport, BundleFlowTreeViewport, DeclarerPromptViewport, ExplanationViewport,
    ModuleCatalogEntry, ModuleFlowTreeViewport, ModuleLearningViewport, PlayCardResult,
    PlayingViewport, SingleCardResult,
};

use crate::error::ServiceError;
use crate::request_types::{SessionConfig, SessionHandle};
use crate::response_types::{
    BidSubmitResult, ConventionInfo, DDSolutionResult, DrillStartResult, InferenceTimelineEntryDTO,
    PromptAcceptResult, ServiceDebugLogEntryDTO, ServiceDebugSnapshotDTO, ServicePublicBeliefState,
};

/// Production service interface — all methods synchronous.
pub trait ServicePort {
    // ── Session lifecycle ──────────────────────────────────────────

    /// Create a new drill session from configuration.
    fn create_session(&mut self, config: SessionConfig) -> Result<SessionHandle, ServiceError>;

    /// Start the drill: generate deal, run initial AI bids, return viewport.
    fn start_drill(&mut self, handle: &str) -> Result<DrillStartResult, ServiceError>;

    // ── Bidding ────────────────────────────────────────────────────

    /// Submit a user bid, get grading + AI continuation.
    fn submit_bid(&mut self, handle: &str, call: Call) -> Result<BidSubmitResult, ServiceError>;

    // ── Phase transitions ──────────────────────────────────────────

    /// Accept a prompt (play/skip/replay/restart).
    fn accept_prompt(
        &mut self,
        handle: &str,
        mode: Option<&str>,
        seat_override: Option<Seat>,
    ) -> Result<PromptAcceptResult, ServiceError>;

    // ── Play ───────────────────────────────────────────────────────

    /// Play a card during the play phase.
    fn play_card(
        &mut self,
        handle: &str,
        card: Card,
        seat: Seat,
    ) -> Result<PlayCardResult, ServiceError>;

    /// Play a single card without running the AI loop.
    /// Used by MC+DDS profiles where TS drives AI card selection.
    fn play_single_card(
        &mut self,
        handle: &str,
        card: Card,
        seat: Seat,
    ) -> Result<SingleCardResult, ServiceError>;

    /// Skip play phase, go directly to review.
    fn skip_to_review(&mut self, handle: &str) -> Result<(), ServiceError>;

    /// Update the play profile for AI play style.
    fn update_play_profile(&mut self, handle: &str, profile_id: &str) -> Result<(), ServiceError>;

    // ── Query (viewport getters) ───────────────────────────────────

    fn get_bidding_viewport(&self, handle: &str) -> Result<Option<BiddingViewport>, ServiceError>;
    fn get_declarer_prompt_viewport(
        &self,
        handle: &str,
    ) -> Result<Option<DeclarerPromptViewport>, ServiceError>;
    fn get_playing_viewport(&self, handle: &str) -> Result<Option<PlayingViewport>, ServiceError>;
    fn get_explanation_viewport(
        &self,
        handle: &str,
    ) -> Result<Option<ExplanationViewport>, ServiceError>;

    // ── Inference ──────────────────────────────────────────────────

    fn get_public_belief_state(
        &self,
        handle: &str,
    ) -> Result<ServicePublicBeliefState, ServiceError>;

    // ── DDS ────────────────────────────────────────────────────────

    fn get_dds_solution(&self, handle: &str) -> Result<DDSolutionResult, ServiceError>;

    /// Get the deal in PBN format for browser DDS solving.
    fn get_deal_pbn(&self, handle: &str) -> Result<String, ServiceError>;

    // ── Catalog ────────────────────────────────────────────────────

    fn list_conventions(&self) -> Vec<ConventionInfo>;
    fn list_modules(&self) -> Vec<ModuleCatalogEntry>;

    // ── Learning ───────────────────────────────────────────────────

    fn get_module_learning_viewport(&self, module_id: &str) -> Option<ModuleLearningViewport>;
    fn get_bundle_flow_tree(&self, bundle_id: &str) -> Option<BundleFlowTreeViewport>;
    fn get_module_flow_tree(&self, module_id: &str) -> Option<ModuleFlowTreeViewport>;
}

/// Debug service methods — separate trait for feature-gating.
pub trait DevServicePort: ServicePort {
    /// Get the expected bid from convention strategy.
    fn get_expected_bid(&self, handle: &str) -> Result<Option<Call>, ServiceError>;

    /// Get a debug snapshot of the current session state.
    fn get_debug_snapshot(&self, handle: &str) -> Result<ServiceDebugSnapshotDTO, ServiceError>;

    /// Get the debug log (list of events).
    fn get_debug_log(&self, handle: &str) -> Result<Vec<ServiceDebugLogEntryDTO>, ServiceError>;

    /// Get the inference timeline.
    fn get_inference_timeline(
        &self,
        handle: &str,
    ) -> Result<Vec<InferenceTimelineEntryDTO>, ServiceError>;

    /// Get the convention name for the active session.
    fn get_convention_name(&self, handle: &str) -> Result<String, ServiceError>;
}
